lib/src/st2110/pipeline/st40_pipeline_tx.c

```diff
@@ -0,0 +1,563 @@
+
+  if (ctx->framebuffs && mt_started(impl)) {
+    tx_st40p_framebuffs_flush(ctx);
+  }
+
+  if (ctx->ready) {
+    mt_stat_unregister(impl, tx_st40p_stat, ctx);
+  }
+
+  if (ctx->transport) {
+    st40_tx_free(ctx->transport);
+    ctx->transport = NULL;
+  }
+  tx_st40p_uinit_fbs(ctx);
+
+  mt_pthread_mutex_destroy(&ctx->lock);
+  mt_pthread_mutex_destroy(&ctx->block_wake_mutex);
+  mt_pthread_cond_destroy(&ctx->block_wake_cond);
+  notice("%s(%d), succ\n", __func__, ctx->idx);
+  mt_rte_free(ctx);
+
+  return 0;
+}
+
+st40p_tx_handle st40p_tx_create(mtl_handle mt, struct st40p_tx_ops* ops) {
+  static int st40p_tx_idx;
+  struct mtl_main_impl* impl = mt;
+  struct st40p_tx_ctx* ctx;
+  int ret;
+  int idx = st40p_tx_idx;
+
+  notice("%s, start for %s\n", __func__, mt_string_safe(ops->name));
+
+  if (impl->type != MT_HANDLE_MAIN) {
+    err("%s, invalid type %d\n", __func__, impl->type);
+    return NULL;
+  }
+
+  enum mtl_port port = mt_port_by_name(impl, ops->port.port[MTL_SESSION_PORT_P]);
+  if (port >= MTL_PORT_MAX) return NULL;
+  int socket = mt_socket_id(impl, port);
+
+  if (ops->flags & ST40P_TX_FLAG_FORCE_NUMA) {
+    err("%s(%d), force numa not supported\n", __func__, idx);
+    return NULL;
+  }
+
+  ctx = mt_rte_zmalloc_socket(sizeof(*ctx), socket);
+  if (!ctx) {
+    err("%s, ctx malloc fail on socket %d\n", __func__, socket);
+    return NULL;
+  }
+
+  ctx->idx = idx;
+  ctx->socket_id = socket;
+  ctx->ready = false;
+  ctx->impl = impl;
+  ctx->type = MT_ST40_HANDLE_PIPELINE_TX;
+  mt_pthread_mutex_init(&ctx->lock, NULL);
+
+  mt_pthread_mutex_init(&ctx->block_wake_mutex, NULL);
+  mt_pthread_cond_wait_init(&ctx->block_wake_cond);
+  ctx->block_timeout_ns = NS_PER_S;
+  if (ops->flags & ST40P_TX_FLAG_BLOCK_GET) {
+    ctx->block_get = true;
+  }
+
+  /* copy ops */
+  if (ops->name) {
+    snprintf(ctx->ops_name, sizeof(ctx->ops_name), "%s", ops->name);
+  } else {
+    snprintf(ctx->ops_name, sizeof(ctx->ops_name), "ST40P_TX_%d", idx);
+  }
+  ctx->ops = *ops;
+
+  ctx->framebuff_cnt = ops->framebuff_cnt;
+  /* init fbs */
+  ret = tx_st40p_init_fbs(ctx, ops);
+  if (ret < 0) {
+    err("%s(%d), init fbs fail %d\n", __func__, idx, ret);
+    st40p_tx_free(ctx);
+    return NULL;
+  }
+
+  /* crete transport handle */
+  ret = tx_st40p_create_transport(impl, ctx, ops);
+  if (ret < 0) {
+    err("%s(%d), create transport fail\n", __func__, idx);
+    st40p_tx_free(ctx);
+    return NULL;
+  }
+
+  /* all ready now */
+  ctx->ready = true;
+  notice("%s(%d), flags 0x%x\n", __func__, idx, ops->flags);
+  st40p_tx_idx++;
+
+  /* notify app can get frame */
+  if (!ctx->block_get) tx_st40p_notify_frame_available(ctx);
+
+  mt_stat_register(impl, tx_st40p_stat, ctx, ctx->ops_name);
+
+  return ctx;
+}
+
+int st40p_tx_update_destination(st40p_tx_handle handle, struct st_tx_dest_info* dst) {
+  struct st40p_tx_ctx* ctx = handle;
+  int cidx = ctx->idx;
+
+  if (ctx->type != MT_ST40_HANDLE_PIPELINE_TX) {
```

Comment:
lets do a switcharoo ok ? 
```suggestion
  if (MT_ST40_HANDLE_PIPELINE_TX != ctx->type) {
```

Reply:
This one, I cannot agree with. It is: if type (variable) is not as set (constant).

Reply:
well i do seem to be at a disadvantage here but i love the 
(constant) != variable 
It nullifies the errors where variable = constant instead of variable == constant 

But my enthusiasm to this is not shared and i will reluctantly fall in line

Reply:
Hi he went and reordered it my way 
GLORY TO THE
CONSTANT == VARIABLE 

@MateuszGrabuszynski  
![image](https://github.com/user-attachments/assets/dc4d5c2f-47ed-448c-8d29-5d4e7b064ab0)